### PR TITLE
Improve handling non-default collection documents rendering and writing

### LINKED ISSUE #5953: Collection documents with future date and output set to false are not ignored
- [x] I believe this to be a bug, not a question about using Jekyll.
- [x] I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
- [x] I read the CONTRIBUTION file at https://jekyllrb.com/docs/contributing/
- [ ] This is a feature request.

---

- [ ] I am on (or have tested on) ***macOS*** 10+
- [x] I am on (or have tested on) ***Debian/Ubuntu*** GNU/Linux (Linux Mint)
- [ ] I am on (or have tested on) ***Fedora*** GNU/Linux
- [ ] I am on (or have tested on) ***Arch*** GNU/Linux
- [x] I am on (or have tested on) ***Other*** GNU/Linux
- [ ] I am on (or have tested on) ***Windows*** 10+

---

- [ ] I was trying to install.
- [ ] There is a broken Plugin API.
- [ ] I had an error on GitHub Pages, and I have reproduced it locally.
- [ ] I had an error on GitHub Pages, and GitHub Support said it was a Jekyll Bug.
- [ ] I had an error on GitHub Pages and I did not test it locally.
- [x] I was trying to build.
- [x] It was another bug.

## My Reproduction Steps

1. I created a collection, we'll call it "mycoll", with the following configuration in `_config.yml`:

```
collections:
  mycoll:
    output: false
```

2. I added a few files to the `_mycoll` directory; one in the past, one in the present, and one in the future. Each one contains front matter similar to the following:

```
date: 2017-03-16 00:00:00-04
link: http://example.com/place
preview: example.png
```

3. I used the following to fetch the latest item on my page:

```
{% assign item = site.mycoll.last %}
```

This is where the problem begins. For posts and collections with `output` set to `true`, posts and documents with the `date` field set to a time in the future are not published by default, unless built with `--future` or `future: true` in the config file; this is documented behavior. However, if `output` is set to `false`, objects with future dates are still displayed. This is not documented behavior.

This behavior is reproducible on both Jekyll 3.4.2 and 3.3.0.

## The Output I Wanted

What I would expect to happen is that ALL collection documents with future dates would be ignored, not just collections with `output: true`. Either that, or the documentation should be updated to reflect the current behavior.

Build output:

```
Configuration file: _config.yml
Configuration file: _config_beta.yml
            Source: _site
       Destination: public
 Incremental build: disabled. Enable with --incremental
      Generating...
Generating Compass: /home/site/beta/_site/_compass => /home/site/beta/public/css
```

/cc @jekyll/build @jekyll/documentation

### PR DESCRIPTION
**Pull request by ashmaroli**
Resolves #5953 
Replaces #6506 

- [x] implement patch
- [x] update cucumber feature and unit test
- [x] update documentation

### CODE REVIEW

File: features/collections.feature [hunk 1]
```diff
@@ -116,7 +116,31 @@ Feature: Collections
...
+    """
+    collections:
+      puppies:
+        output: true
+    """
+    And I have a "index.html" page that contains "Newest puppy: {% assign puppy = site.puppies.last %}{{ puppy.title }}"
+    When I run jekyll build
+    Then I should get a zero exit status
+    And the _site directory should exist
+    And I should see "Newest puppy: Fido" in "_site/index.html"
```
Comment (DirtyF) [inline features/collections.feature]: Why do we return a future document here?

File: features/collections.feature [hunk 2]
```diff
@@ -139,6 +163,30 @@ Feature: Collections
...
+    """
+    collections:
+      puppies:
+        output: false
+    """
+    And I have a "index.html" page that contains "Newest puppy: {% assign puppy = site.puppies.last %}{{ puppy.title }}"
+    When I run jekyll build
+    Then I should get a zero exit status
+    And the _site directory should exist
+    And I should see "Newest puppy: Fido" in "_site/index.html"
```
Comment (DirtyF) [inline features/collections.feature]: Same question as above.

File: features/collections.feature (hunk 1 above)
Comment (ashmaroli) [reply features/collections.feature]: I've documented the behavior in `_docs/collections.md`

This is based on https://github.com/jekyll/jekyll/pull/6506#issuecomment-340670610 and  https://github.com/jekyll/jekyll/pull/6506#issuecomment-340675231 by @pathawks.
You can iterate through *future docs* but won't be output / written to disk

Comment (ashmaroli) [reply features/collections.feature]: If this a breaking-change in behavior, we can ship this in `4.0`.. no biggie

Comment (DirtyF) [reply features/collections.feature]: @pathawks use case with future events makes sense. 
Hugo's `date`, `publishsate` and `expirydate` is way more flexible.

Comment (ashmaroli) [reply features/collections.feature]: If Pat Hawks use case is possible on `master` then this PR is safe to ship in `3.x` otherwise its probably a `4.0`

Comment (pathawks) [reply features/collections.feature]: On `master`, future docs are rendered and can be accessed, but are not written to disk. :+1:

File: lib/jekyll/collection.rb [hunk 3]
```diff
@@ -210,12 +210,11 @@ def container
...
-      doc = Jekyll::Document.new(full_path, :site => site, :collection => self)
-      doc.read
-      if site.publisher.publish?(doc) || !write?
-        docs << doc
-      else
-        Jekyll.logger.debug "Skipped Publishing:", doc.relative_path
+      docs << Document.new(full_path, :site => site, :collection => self).tap do |doc|
+        doc.read
+        if !site.publisher.publish?(doc) && site.publisher.hidden_in_the_future?(doc)
+          Jekyll.logger.debug "Skip Publishing:", "#{doc.relative_path} has a future date"
```
Comment (pathawks) [inline lib/jekyll/collection.rb]: I am unable to trigger this log message. Perhaps I am misunderstanding.

https://github.com/jekyll/cases/commit/0c7256b03a1c24bf9afe2ad3c45627a1e3fabd1b

Comment (ashmaroli) [reply lib/jekyll/collection.rb]: can you post the *entire* verbose build output under the commit? Thanks

Comment (pathawks) [reply lib/jekyll/collection.rb]: **verbose** is the key word. 🤕

File: docs/_docs/collections.md [hunk 4]
```diff
@@ -348,6 +348,20 @@ you specified in your `_config.yml` (if present) and the following information:
   <code>{% raw %}{{ site.collections | where: "label", "myCollection" | first }}{% endraw %}</code></p>
 </div>
 
+<div class="note info">
+  <h5>Collections and Time</h5>
+  <p>Except for documents in hard-coded default collection <code>posts</code>, all documents in collections
+    you create, are accessible via Liquid irrespective of their assigned date, if any, and therefore renderable.
+  </p>
+  <p>However documents are attempted to be physically output (written to disk) only if the concerned collection
```
Comment (oe) [inline docs/_docs/collections.md]: just say written to disk here

File: docs/_docs/collections.md [hunk 5]
```diff
@@ -348,6 +348,20 @@ you specified in your `_config.yml` (if present) and the following information:
...
+    metadata has <code>output: true</code>. Additionally, future-dated documents are only written if
+    <code>site.future</code> <em>is also true</em>.
+  </p>
+  <p>A more fine-grained control over documents being written to disk can be further excercised by setting
```
Comment (oe) [inline docs/_docs/collections.md]: `More fine-grained control [...]`

**Comment by oe** [comment]
The title is kind of misleading, no? There's more than that added in this PR.

**Comment by ashmaroli** [comment]
> The title is kind of misleading, no?

Feel free to change the title to something that fits the proposed changes.

**Comment by ashmaroli** [comment]
```
Improve handling non-default collection documents rendering and writing
```

**Comment by DirtyF** [comment]
…waiting for @oe approval before merging this fix.

**Comment by DirtyF** [comment]
@jekyllbot: :shipit: +fix

**Pull request merged**